extensions/script/python/ExecutePythonProcessor.cpp

```diff
@@ -35,155 +35,184 @@ namespace python {
+    engine->onTrigger(context, session);
+    handleEngineNoLongerInUse(std::move(engine));
+  }
+  catch (const std::exception &exception) {
+    logger_->log_error("Caught Exception: %s", exception.what());
     this->yield();
-  } catch (...) {
+  }
+  catch (...) {
     logger_->log_error("Caught Exception");
     this->yield();
   }
 }
 
+// TODO(hunyadi): This is potentially not what we want. See https://issues.apache.org/jira/browse/MINIFICPP-1222
+std::shared_ptr<python::PythonScriptEngine> ExecutePythonProcessor::getScriptEngine() {
+  std::shared_ptr<python::PythonScriptEngine> engine;
+  // Use an existing engine, if one is available
+  if (script_engine_q_.try_dequeue(engine)) {
+    logger_->log_debug("Using available [%p] script engine instance", engine.get());
+    return engine;
+  }
+  engine = createEngine<python::PythonScriptEngine>();
+  logger_->log_info("Created new [%p] script engine instance. Number of instances: approx. %d / %d.", engine.get(), script_engine_q_.size_approx(), getMaxConcurrentTasks());
+  if (engine == nullptr) {
+    throw std::runtime_error("No script engine available");
+  }
+  return engine;
+}
+
+void ExecutePythonProcessor::handleEngineNoLongerInUse(std::shared_ptr<python::PythonScriptEngine>&& engine) {
+  // Make engine available for use again
+  if (script_engine_q_.size_approx() < getMaxConcurrentTasks()) {
+    logger_->log_debug("Releasing [%p] script engine", engine.get());
+    script_engine_q_.enqueue(engine);
+  } else {
+    logger_->log_info("Destroying script engine because it is no longer needed");
+  }
+}
+
+void ExecutePythonProcessor::appendPathForImportModules() {
+  // TODO(hunyadi): I have spent some time trying to figure out pybind11, but
+  // could not get this working yet. It is up to be implemented later
+  // https://issues.apache.org/jira/browse/MINIFICPP-1224
+  if (module_directory_.size()) {
+    logger_->log_error("Not supported property: Module Directory.");
+  }
+}
+
+void ExecutePythonProcessor::loadScriptFromFile(const std::string& file_path) {
+  std::ifstream file_handle(file_path);
+  if (!file_handle.is_open()) {
+    script_to_exec_ = "";
+    throw std::runtime_error("Failed to read Script File: " + file_path);
+  }
+  script_to_exec_ = std::string{ (std::istreambuf_iterator<char>(file_handle)), (std::istreambuf_iterator<char>()) };
+}
+
+void ExecutePythonProcessor::loadScript() {
+  std::string script_file;
+  std::string script_body;
+  getProperty(ScriptFile.getName(), script_file);
+  getProperty(ScriptBody.getName(), script_body);
+  if (script_file.size()) {
+    if (script_body.size()) {
+      throw std::runtime_error("Only one of Script File or Script Body may be used");
+    }
+    loadScriptFromFile(script_file);
+    return;
+  }
+  if (script_body.size()) {
+    script_to_exec_ = script_body;
+    return;
+  }
+  throw std::runtime_error("Neither Script Body nor Script File is available to execute");
```

Comment:
Not a big deal, but I prefer a structure of the function that throws in error branches.

```
if (script_body.size() == 0) {
  throw std::runtime_error("Neither Script Body nor Script File is available to execute");
}
script_to_exec_ = script_body;
```

This way it seems to be shorter and better structured to me.

Reply:
Restructured.